src/csharp/Grpc.Core/Marshaller.cs

```diff
@@ -29,36 +29,122 @@ public class Marshaller<T>
         readonly Func<T, byte[]> serializer;
         readonly Func<byte[], T> deserializer;
 
+        readonly Action<T, SerializationContext> contextualSerializer;
+        readonly Func<DeserializationContext, T> contextualDeserializer;
+
         /// <summary>
-        /// Initializes a new marshaller.
+        /// Initializes a new marshaller from simple serialize/deserialize functions.
         /// </summary>
         /// <param name="serializer">Function that will be used to serialize messages.</param>
         /// <param name="deserializer">Function that will be used to deserialize messages.</param>
         public Marshaller(Func<T, byte[]> serializer, Func<byte[], T> deserializer)
         {
-            this.serializer = GrpcPreconditions.CheckNotNull(serializer, "serializer");
-            this.deserializer = GrpcPreconditions.CheckNotNull(deserializer, "deserializer");
+            this.serializer = GrpcPreconditions.CheckNotNull(serializer, nameof(serializer));
+            this.deserializer = GrpcPreconditions.CheckNotNull(deserializer, nameof(deserializer));
+            this.contextualSerializer = EmulateContextualSerializer;
```

Comment:
When this "old-style" constructor is used, when will `this.contextualSerializer` and `this.contextualDeserializer` ever be used by the gRPC-C# library? Is the plan to switch over gRPC-C# to start calling the `contextualSerializer` and `contextualDelegates`?

Reply:
Yes, the plan is to switch gRPC C# library to use contextualSerializer and contextualDeserializer because they are more general and their implementation can trivially invoke the old-style serializer and deserializer with no extra overhead (I plan to do the switch in #16371).